### ISSUE TITLE
Stubbing Input Class

### ISSUE THREAD

**Issue by myin142** [bitwes/Gut#171]
Is there any way to stub the return values of the Input class? e.g `Input.is_action_just_pressed`

In the example tutorials it seems like they were just not tested. Is it not possible or should it just not be tested?

**Comment by bitwes** [comment]
I haven't tested all of the native classes, so some might not work.

I tried to `double` and `partial_double` the `Input` and `InputDefault` classes and couldn't get it to work.  It appears that `Input` uses `InputDefault` somehow.  I tried creating an intermediate class that `extends Input` or `extends InputDefault` and then doubling an instance of that but that wouldn't work either.  I then just tried to create vanilla instance with `Input.new()` which errors out.  It looks like you can do `InputDefault.new()` but I couldn't get that to work as a double.

The documentation says the `Input` class is "A singleton that deals with inputs.".  So this might be why it can't be doubled.

You might be able to work around this if you create a class that has all the methods/properties of `Input` that you use in your functions.  You shouldn't have to make anything other than what you use.  You can then stub an instance of that class and pass it around.  You just can't get all the functions for free like you do with a double.

**Comment by bitwes** [comment]
A fix to this might be that native class doubles (`Input`, `File` etc) shouldn't extend their source class.  The code already creates overloads for all the methods that the native class has.  It doesn't create any properties though, so that would have to be added.  If you did this, then you couldn't `partial_double` them since there would be no super class to punch through to.  It also means that doubles won't pass an `is` check which could be problematic.  Maybe doing it this way could be an additional form of doubling.  Instead of the tool knowing which kind to use it could be left to the user.  It could be called `double_without_inherit` 

Doubling native classes is a messy endeavor, but being able to stub something like `Input` seems important.  You can of course get most of it tested with modularization and the method mentioned above, but this would be great to figure out.  Just not sure if there is a good solution.

**Comment by Xrayez** [comment]
Note that `InputDefault` functionality is merged into a single `Input` singleton in development version (Godot 4.0).

Regarding the "stubbing return values", I think it would make more sense to do this if Godot had something akin to dedicated `InputState` or `InputController` classes (implemented as a resource within the `Input` singleton), which you could then double I suppose. See my proposal at godotengine/godot-proposals#104.

**Comment by nightblade9** [comment]
I found a rather easy/simple work-around for this. Inspired by C#, you can use manual dependency injection to solve this.

Let's say you have a `Player` class that depends on `Input.is_action_pressed` to set velocity. Here's how you can make it accept a stub/mock input class, while defaulting to the Godot "production" `Input` class:

```
# Player.gd
var _input:Object = Input; # default to Godot's Input class
def initialize(input):
  _input = input

def _process_physics(delta):
  if _input.is_action_pressed(...) # as usual
# ...
```

In your test, pass in a mock/stub/whatever:

```
# class: test_player.gd
class MockInput:
  var pressed = []
  func press(key):
    _pressed.push(key)
 
  func release(key):
    _pressed.remove(key)

  func is_action_pressed(a):
    return a in pressed

# actual test
func test_stuff():
  var p = Player.new() # Player.instance() works also
  var m = MockInput.new()
  m.press("ui_right")
  # ... assert player moved right ... 
```

In your actual game, calling `Player.instance()` will get an instance with the default (wired up to Godot).

**Comment by bitwes** [comment]
This is a pretty good workaround.  Nice.  Do you have to default the constructor parameter or is there some c# magic going on?  This still requires some manual work on the `MockInput` class to get all the functionality of `Input` but  this solves 1/2 the problem and will be great in cases where you aren't checking a lot  of different kinds of input.

For the non-c#ers, the gdscript would look like this:
```
# Player.gd
var _input = Input # default to Godot's Input class
func _init(input = Input):
  _input = input

func _process_physics(delta):
  if _input.is_action_pressed(...) # as usual
# ...
```
and  then this (in both cases)
```
func test_stuff():
  var m = MockInput.new()
  var p = Player.new(m) # Player.instance() works also
  m.press("ui_right")
```
Or add accessors if it gets messy with your constructor.

**Comment by bitwes** [comment]
`send_keys` popped into my head and I found there is a  way to send input events to `Input`.  Haven't tried it yet.
https://godotengine.org/qa/5896/simulate-mouse-and-keyboard-events-by-code

**Comment by nightblade9** [comment]
> . Do you have to default the constructor parameter or is there some c# magic going on?

This is all GDscript, there's no C#. It's inspired by how I would handle this problem in C#.

The reason I have an `initialize` method and not `_init`, is because calling `Foo.instance()` doesn't call `_init`. Having your own method (I like to call mine `initialize` isn't as great since you have to manually call it after construction, but it's a nice way to handle this.

I'm not defaulting the constructor parameter, because in `Player.gd`, the default is already set to `Input`:

```
# Player.gd
extends Node2D
  var _input:Object = Input
```

No it's not a complete mock/stub, but it's enough of a "proof of concept" - I am just using `is_action_pressed`, you can add functionality for anything else in `Input` that you want to mock.

**Comment by bitwes** [comment]
Ah, ok.  The mix of `def` and `func` threw me off.  It didn't look like c# but I thought maybe it was  CSharpScript, whatever that is. 

In GDScript, `_init` is called when `instance` is called.  I'm not sure if that's the case for `GDNative` classes.  

"I'm not defaulting the constructor parameter":  Ok, I was confused since you were making a `Player` instance and not sending it the `MockInput` and without a default `new` requires something to be passed in.

Maybe a first step would be to create something in GUT that  mocks out Input that you could then use with this approach.

**Comment by nightblade9** [comment]
My bad. I wrote that Python-style by mistake (it's fixed now). `class` is apparently a GDscript keyword you can use to make an inner class.

Are you asking me to contribute something to GUT? Sure, but can you clarify what exactly you want me to do? (I have this working in my own project, with code similar to the pseudocode I posted above.)

**Comment by bitwes** [comment]
I wasn't asking you specifically, just stating that might be good first step.  Also, I'm not really sure what it would look like.  I see two approaches off the top of my head.

# Provide a stubbable (fun to say) `Input` class.
* Find  a way to create a class that has all the functionality of `Input` but does nothing.  We'll call it `MockInput` for now.
  * This could  be done by hand but that seems like a bad idea  for maintenance.
  * Implementing the  ability to create a double that does not extend the  base class would probably  be the way to go here.  Given  there might not be a lot of use  cases for this it should be private at first.
* Add a method to `addons/gut/test.gd` that will return an instance of `MockInput`
* You would then stub this instance for your test and pass it off to your class so it can use it instead of `Input`.

The downside with this approach is that you have to make you class  a specific way to test it.  It will  also be difficult to  simulate multiple actions in a row.  If you wanted to test a cheat code for example you are going to need 2 lines per keypress (one for the stub and one to process the input).  

# Create methods to simulate mouse and keyboard input.
As outlined in [this post](https://godotengine.org/qa/5896/simulate-mouse-and-keyboard-events-by-code) it seems you can  send input to `Input`.  Creating some convenience wrappers for these could be enough.  This would probably have to go into a utility class with a bunch of static methods instead of `test.gd`.  Call it something like `PhyscialInteractionDeviceSimulatorSupreme`.
* `send_key_pressed`
* `send_key_down`
* `send_key_up`
* `send_click`
* `send_mouse_down`
* `send_mouse_up`
* `move_mouse_to`
* `send_action`  (for `Input.is_action_pressed("ui_right")`)
*  probably 10 more I haven't thought of yet.

The downside here is that I can see it getting ugly to test in complex situations, this is sending input the global `Input` instance.  The inputs might be interpreted by something else if you are not careful.  It's also a lot of methods to make.

**Comment by bitwes** [comment]
@myin142 mentions `InputEventAction` in #250.  Is this useful here?

**Comment by myin142** [comment]
I have changed my structure of how I handle the inputs in my games where I don't have this problem anymore. AFAIK it is better to get the input from `_unhandled_input`, so I have changed all the classes to only get input from there.
Now I can easily mock the input by creating an InputEvent and pass it to `get_tree().input_event()`. The class `InputEventAction` is just an easy way to create an InputEvent that I found. Here a simple example: (input is the name of the key e.g "ui_select")

```gdscript
var ev = InputEventAction.new()
ev.action = input
ev.pressed = pressed
ev.strength = strength
```

**Comment by bitwes** [comment]
I've updated the approach comment I had made earlier.  I've figured out a way to `double(Input)` which should prove useful.  Also added details on a new helper class `InputSender`.

**Comment by bitwes** [comment]
7.3.0 will `InputFactory`, `InputSender` and the ability to create stubs of Godot Singletons such as `Input`, `OS` and `Javascript` which address all the features in the [comment above](https://github.com/bitwes/Gut/issues/171#issuecomment-700254936).

These changes are currently in master.

**Comment by bitwes** [comment]
I ran into some issues doubling singletons and created issue #346 to capture that effort.  `InputFactory` and `InputSender` have been released in v7.3.0.
